checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallInference.java

```diff
@@ -1,176 +1,769 @@
+      inferOwningParamsViaOwnershipTransfer(obligations, invocation);
+    } else if (invocation instanceof MethodInvocationNode) {
+      inferOwningParamsViaOwnershipTransfer(obligations, invocation);
+      computeOwningForArgsOrReceiverOfCall(obligations, (MethodInvocationNode) invocation);
+    }
+  }
+
+  /**
+   * Returns the called methods annotation for the given Java expression after the invocation node.
+   *
+   * @param invocation the MethodInvocationNode
+   * @param varJe a Java expression
+   * @return the called methods annotation for the {@code varJe} after the {@code invocation} node.
+   */
+  private AnnotationMirror getCalledMethodsAnno(
+      MethodInvocationNode invocation, JavaExpression varJe) {
+    AccumulationStore cmStoreAfter = resourceLeakAtf.getStoreAfter(invocation);
+    AccumulationValue cmValue = cmStoreAfter == null ? null : cmStoreAfter.getValue(varJe);
+
+    AnnotationMirror cmAnno = null;
+
+    if (cmValue != null) {
+      // The store contains the lhs.
+      Set<String> accumulatedValues = cmValue.getAccumulatedValues();
+      if (accumulatedValues != null) { // type variable or wildcard type
+        cmAnno = resourceLeakAtf.createCalledMethods(accumulatedValues.toArray(new String[0]));
+      } else {
+        for (AnnotationMirror anno : cmValue.getAnnotations()) {
+          if (AnnotationUtils.areSameByName(
+              anno, "org.checkerframework.checker.calledmethods.qual.CalledMethods")) {
+            cmAnno = anno;
+          }
+        }
+      }
+    }
+
+    if (cmAnno == null) {
+      cmAnno = resourceLeakAtf.top;
+    }
+
+    return cmAnno;
+  }
+
+  /**
+   * Adds all non-exceptional successors to {@code worklist}. If a successor is a non-exceptional
+   * exit point, adds an {@literal @Owning} annotation for fields in {@link #disposedFields}.
+   *
+   * @param obligations the obligations for the current block
+   * @param curBlock the block whose successors to add to the worklist
+   * @param visited block-Obligations pairs already analyzed or already on the worklist
+   * @param worklist the worklist, which is side-effected by this method
+   */
+  private void addNonExceptionalSuccessorsToWorklist(
+      Set<Obligation> obligations,
+      Block curBlock,
+      Set<BlockWithObligations> visited,
+      Deque<BlockWithObligations> worklist) {
+
+    for (Block successor : getNonExceptionalSuccessors(curBlock)) {
+      // If successor is a special block, it must be the regular exit.
+      if (successor.getType() == Block.BlockType.SPECIAL_BLOCK) {
+        WholeProgramInference wpi = resourceLeakAtf.getWholeProgramInference();
+        assert wpi != null : "MustCallInference is running without WPI.";
+        for (VariableElement fieldElt : updateOwningFields()) {
+          wpi.addFieldDeclarationAnnotation(fieldElt, OWNING);
+        }
+        if (!disposedFields.isEmpty()) {
+          addEnsuresCalledMethods();
+        }
+
+        addOrUpdateClassMustCall();
```

Comment:
It seems this code will run _every_ time the regular exit is reached during the analysis.  Do we really want this, or should we run it one time, at the end of `runInference()` after the outer while loop of the analysis has completed?  What could happen here is:

1. We reach the exit block along one path, with field `f` in `disposedFields`, so we add an `@EnsuresCalledMethods` annotation.
2. Along another path, we remove `f` from `disposedFields`, due to an overwrite.  But, it's too late to remove it, as we have already added an `@EnsuresCalledMethods` annotation to the method.

Beyond avoiding the above scenario, it seems more efficient and also makes more sense (to me at least) to complete analysis before adding these annotations.